pubsub/envelope.go

```diff
@@ -55,22 +45,63 @@ func NewCloudEventsEnvelope(id, source, eventType, subject string, data []byte)
 	if eventType == "" {
 		eventType = DefaultCloudEventType
 	}
+	if subject == "" {
+		subject = DefaultCloudEventSource
+	}
+
+	// check if JSON
+	var j interface{}
+	err := jsoniter.Unmarshal(data, &j)
+	if err != nil {
+		// not JSON, return new envelope
+		return &CloudEventsEnvelope{
+			ID:              id,
+			SpecVersion:     CloudEventsSpecVersion,
+			DataContentType: "text/plain",
+			Source:          source,
+			Type:            eventType,
+			Subject:         subject,
+			Data:            string(data),
+		}
+	}
+
+	// handle CloudEvent
+	m, isMap := j.(map[string]interface{})
+	if isMap {
+		if _, isCE := m["specversion"]; isCE {
+			ce := &CloudEventsEnvelope{
+				ID:              getStrVal(m, "id"),
+				SpecVersion:     getStrVal(m, "specversion"),
+				DataContentType: getStrVal(m, "datacontenttype"),
+				Source:          getStrVal(m, "source"),
+				Type:            getStrVal(m, "type"),
+				Subject:         getStrVal(m, "subject"),
+				Data:            m["data"],
+			}
+			// check if CE is valid
+			if ce.ID != "" && ce.SpecVersion != "" && ce.DataContentType != "" {
+				return ce
+			}
+		}
+	}
 
-	// create new envelope
-	ce = CloudEventsEnvelope{
+	// content was JSON but not a valid CloudEvent, make one
+	return &CloudEventsEnvelope{
 		ID:              id,
 		SpecVersion:     CloudEventsSpecVersion,
 		DataContentType: "application/json",
 		Source:          source,
 		Type:            eventType,
 		Subject:         subject,
+		Data:            string(data),
```

Comment:
This must be `j` which is the deserialized object. Otherwise, data will be treated as normal string in serialization.

please see the original code where it took the deserialized object `i`.

Reply:
https://github.com/dapr/components-contrib/pull/389/files#r453886658 <- original code

Reply:
This make sense. Made that change but had to change our existing unit tests. 

We are now testing for: 

* JSON obj
* JSON string
* Non-JSON content 
* Existing CloudEvent

PTAL